java/src/main/native/src/row_conversion.cu

```diff
@@ -304,8 +329,620 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
+      // to row it is much easier to copy by elements first otherwise we would need a running
+      // total of the column sizes for our block, which isn't readily available. This makes it
+      // more appealing to copy element-wise from input data into shared matching the end layout
+      // and do row-based memcopies out.
+
+      auto const shared_buffer_base = shared[fetch % stages_count];
+      for (auto el = (int)threadIdx.x; el < num_elements_in_block; el += blockDim.x) {
+        auto const relative_col = el / num_fetch_rows;
+        auto const relative_row = el % num_fetch_rows;
+        auto const absolute_col = relative_col + fetch_block.start_col;
+        auto const absolute_row = relative_row + fetch_block.start_row;
+        auto const col_size = col_sizes[absolute_col];
+        auto const col_offset = col_offsets[absolute_col];
+        auto const relative_col_offset = col_offset - starting_column_offset;
+
+        auto const shared_offset = relative_row * fetch_block_row_size + relative_col_offset;
+        auto const input_src = input_data[absolute_col] + col_size * absolute_row;
+
+        // copy the element from global memory
+        switch (col_size) {
+          case 2:
+            cuda::memcpy_async(&shared_buffer_base[shared_offset], input_src,
+                               cuda::aligned_size_t<2>(col_size), fetch_barrier);
+            break;
+          case 4:
+            cuda::memcpy_async(&shared_buffer_base[shared_offset], input_src,
+                               cuda::aligned_size_t<4>(col_size), fetch_barrier);
+            break;
+          case 8:
+            cuda::memcpy_async(&shared_buffer_base[shared_offset], input_src,
+                               cuda::aligned_size_t<8>(col_size), fetch_barrier);
+            break;
+          default:
+            cuda::memcpy_async(&shared_buffer_base[shared_offset], input_src, col_size,
+                               fetch_barrier);
+            break;
+        }
+      }
+    }
+
+    auto &subset_barrier = block_barrier[subset % NUM_BLOCKS_PER_KERNEL_LOADED];
+    subset_barrier.arrive_and_wait();
+
+    auto block = block_infos[blockIdx.x * NUM_BLOCKS_PER_KERNEL_TO_ROWS + subset];
+    auto const block_row_size = block.get_shared_row_size(col_offsets, col_sizes);
+    auto const column_offset = col_offsets[block.start_col];
+    auto const block_output_buffer = output_data[block.batch_number];
+
+    // copy entire rows to final dest
+    for (auto absolute_row = block.start_row + threadIdx.x; absolute_row <= block.end_row;
+         absolute_row += blockDim.x) {
+      auto const relative_row = absolute_row - block.start_row;
+      auto const output_dest = block_output_buffer + row_offsets[absolute_row] + column_offset;
+      auto const shared_offset = block_row_size * relative_row;
+
+      cuda::memcpy_async(output_dest, &shared[subset % stages_count][shared_offset],
```

Comment:
This `cuda::memcpy_async` doesn't have a similar switch statement as above. (line 451). I assume we expect rows to have more than 8 bytes in size?

Reply:
Yes, rows are padded to 8 bytes for the JCUDF format, so when we're coping out to the final destination we can write in large chunks.